File: frontend/src/types/localPr.ts
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Frontend mirror of the backend local-PR data contract (see the LocalPR
 * domain records + their DTOs). A "local PR" is the full pull-request
 * artifact — description, commits, timeline, checks, comments — living in
 * ByteQuay before it is ever pushed to GitHub. The unified `<PRView mode>`
 * renders both the local phase (`mode="local"`) and the remote phase
 * (`mode="remote"`) from these shapes; `mode` switches the data source and a
 * few affordances, not the visual template.
 */

export type LocalPRStatus =
  | 'local-drafted' // agent writing commits + description
  | 'local-open' // dev done, awaiting user review
  | 'remote-drafted' // pushed as Draft on GitHub
  | 'remote-open' // ready-for-review on GitHub
  | 'merged'
  | 'closed';

/** `origin=task` — created by the dev agent, full local→remote lifecycle.
 *  `origin=external` — discovered via the dashboard sync (someone else's PR,
 *  or our own PR opened outside ByteQuay); never occupies the local-only
 *  statuses. See `derivePRCapabilities`. */
export type PROrigin = 'task' | 'external';

export interface LocalPR {
  id: string;
  taskId: string | null; // null for origin=external
  branchName: string;
  baseBranch: string;
  title: string;
  description: string; // markdown
  status: LocalPRStatus;
  createdAt: number; // epoch ms
  pushedAt: number | null;
  remotePrNumber: number | null;
  remotePrUrl: string | null;
  mergedAt: number | null;
  closedAt: number | null;
  origin: PROrigin;
  repo: string | null; // "owner/name", set for origin=external
  author: string | null; // set for origin=external
  syncedAt: number | null; // last successful GitHub sync, null pre-first-sync
  /** GitHub's own PR-level diff totals (origin=external only) — GitHub's
   *  commit-list API has no per-commit stats, so summing `LocalPRCommit`s
   *  the way a task-origin PR does always reads 0. Null pre-first-sync. */
  syncedAdditions: number | null;
  syncedDeletions: number | null;
  /** GitHub's mergeable/mergeableState for the base branch (origin=external
   *  only) — null until GitHub has computed it (or pre-first-sync). Powers
   *  the merge-box's "No conflicts with base branch" line. */
  syncedMergeable: boolean | null;
  syncedMergeableState: string | null;
  /** True when the PR's base branch has a merge queue configured — GraphQL-
   *  sourced (REST doesn't expose this). Drives the merge-box's "Merge when
   *  ready" button mode instead of a direct method-dropdown merge. */
  syncedMergeQueueEnabled: boolean;
  /** GitHub's per-PR merge-queue entry state ("QUEUED", etc.) once the PR
   *  has joined the queue; null otherwise. */
  syncedMergeQueueState: string | null;
  /** Epoch ms the app deleted the head branch after a merge; null until
   *  the user clicks "Delete branch" (or it's never been clicked). */
  branchDeletedAt: number | null;
}

export interface LocalPRCommit {
  id: string;
  localPrId: string;
  sha: string;
  message: string;
  additions: number;
  deletions: number;
  authoredAt: number;
  pushedAt: number | null;
}

export type LocalPRTimelineEventType =
  | 'commit'
  | 'ci'
  | 'amend'
  | 'branch'
  | 'status'
  | 'review'
  | 'comment'
  | 'follow-up'
  | 'plan-finalized'
  | 'pull-request-progress'
  | 'pull-request-created';

export interface LocalPRTimelineEvent {
  id: string;
  localPrId: string;
  eventType: LocalPRTimelineEventType;
  actor: string; // "claude-code" | "you" | "@<github-user>"
  isLocalOnly: boolean; // stripped on push
  strippedOnPushAt: number | null;
  createdAt: number;
  payload: Record<string, unknown> | null;
  remoteEventId?: number | null;
}

export type LocalPRCheckKind = 'local' | 'remote';
export type LocalPRCheckStatus = 'pending' | 'running' | 'passed' | 'failed' | 'neutral';

export interface LocalPRCheck {
  id: string;
  localPrId: string;
  kind: LocalPRCheckKind;
  name: string;
  status: LocalPRCheckStatus;
  durationMs: number | null;
  startedAt: number;
  finishedAt: number | null;
  runId: string | null;
}

export type LocalPRCommentOrigin = 'local' | 'remote';
export type LocalPRCommentScope = 'pr' | 'file-line';

export interface LocalPRComment {
  id: string;
  localPrId: string;
  origin: LocalPRCommentOrigin;
  scope: LocalPRCommentScope;
  filePath: string | null;
  lineNumber: number | null;
  /** 'LEFT' (removed) or 'RIGHT' (added/context) — defaults to 'RIGHT' for
   *  every comment that predates this concept. */
  side: 'LEFT' | 'RIGHT';
  /** First line of a multi-line range; null for a single-line comment. */
  startLine: number | null;
  /** Diff side of `startLine`; null for a single-line comment. */
  startSide: 'LEFT' | 'RIGHT' | null;
  author: string;
  body: string;
  createdAt: number;
  resolvedAt: number | null;
  dismissedAt: number | null;
  strippedOnPushAt: number | null;
  parentCommentId: string | null;
  /** Set once `publish-review` batches this draft into a GitHub review
   *  (origin=external only — task-origin drafts are stripped on push
   *  instead and never reach this state). */
  publishedAt: number | null;
  /** Investigation finding backing this comment. Null/absent for ordinary
   * human and draft comments. */
  findingId?: string | null;
  /** Actor who resolved the thread ('you' / dev agent id), for the
   *  "X marked this conversation as resolved" attribution. Null while open. */
  resolvedBy?: string | null;
}

/** Everything the `<PRView>` needs, as one bundle a bridge hook resolves. */
export interface LocalPRBundle {
  pr: LocalPR;
  commits: LocalPRCommit[];
  timeline: LocalPRTimelineEvent[];
  checks: LocalPRCheck[];
  comments: LocalPRComment[];
  /** Authoritative count of local-only events + local comments a push would
   *  strip (design #47) — the push dialog shows this verbatim. Optional so
   *  presentational fixtures can omit it. */
  pendingStripCount?: number;
  /** True while the backend is still refreshing this PR from git/GitHub in the
   *  background — the snapshot above is served immediately and may be a beat
   *  behind. `usePR` polls faster until it clears. */
  syncing?: boolean;
}

File: backend/src/main/java/com/bytequay/app/domain/LocalPR.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.bytequay.app.domain;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;

import java.time.Instant;
import java.util.Map;
import java.util.Set;

/**
 * A pull request in its local phase — the full PR artifact (title,
 * description, and its commits / timeline / checks / comments child rows)
 * living entirely in ByteQuay before it ever reaches GitHub. It is pushed
 * to GitHub only on explicit user approval, after which it carries the
 * remote PR number / url and follows the remote half of the state machine.
 *
 * <p>The status wire values match the TypeScript {@code LocalPRStatus} union
 * (hyphenated), so — like {@link BacklogItem} — status is a {@code String}
 * with {@code STATUS_*} constants rather than a Java enum (enum names can't
 * carry hyphens). {@link #ALLOWED_TRANSITIONS} encodes the design's state
 * machine; every accepted flip is validated with {@link #canTransitionTo}.
 * Branch/push/merge runtime state is owned by the task row — {@code branchName}
 * is copied from the task at create time.
 */
public record LocalPR(
        String id,
        String taskId,
        String branchName,
        String baseBranch,
        String title,
        String description,
        String status,
        Instant createdAt,
        Instant pushedAt,
        Integer remotePrNumber,
        String remotePrUrl,
        Instant mergedAt,
        Instant closedAt,
        Instant localAddressedThroughAt)
{
    public static final String STATUS_LOCAL_DRAFTED = "local-drafted";
    public static final String STATUS_LOCAL_OPEN = "local-open";
    public static final String STATUS_REMOTE_DRAFTED = "remote-drafted";
    public static final String STATUS_REMOTE_OPEN = "remote-open";
    public static final String STATUS_MERGED = "merged";
    public static final String STATUS_CLOSED = "closed";

    /**
     * Legal status flips (design #45). A task can be abandoned from any
     * non-terminal state, so every one also permits {@code closed}. {@code
     * merged} and {@code closed} are terminal (no outgoing edges).
     */
    public static final Map<String, Set<String>> ALLOWED_TRANSITIONS = ImmutableMap.of(
            STATUS_LOCAL_DRAFTED, ImmutableSet.of(STATUS_LOCAL_OPEN, STATUS_CLOSED),
            STATUS_LOCAL_OPEN, ImmutableSet.of(STATUS_REMOTE_DRAFTED, STATUS_CLOSED),
            STATUS_REMOTE_DRAFTED, ImmutableSet.of(STATUS_REMOTE_OPEN, STATUS_MERGED, STATUS_CLOSED),
            STATUS_REMOTE_OPEN, ImmutableSet.of(STATUS_MERGED, STATUS_CLOSED),
            STATUS_MERGED, ImmutableSet.of(),
            STATUS_CLOSED, ImmutableSet.of());

    /** A freshly-created local PR: {@code local-drafted}, no push/merge stamps. */
    public static LocalPR create(
            String id,
            String taskId,
            String branchName,
            String baseBranch,
            String title,
            String description,
            Instant createdAt)
    {
        return new LocalPR(
                id, taskId, branchName, baseBranch, title,
                description == null ? "" : description,
                STATUS_LOCAL_DRAFTED, createdAt,
                /* pushedAt */ null, /* remotePrNumber */ null, /* remotePrUrl */ null,
                /* mergedAt */ null, /* closedAt */ null, /* localAddressedThroughAt */ null);
    }

    /** True iff {@code target} is a legal next status from the current one. */
    public boolean canTransitionTo(String target)
    {
        return ALLOWED_TRANSITIONS.getOrDefault(status, Set.of()).contains(target);
    }

    /** Whether this PR has reached a terminal state (merged / closed). */
    public boolean isTerminal()
    {
        return STATUS_MERGED.equals(status) || STATUS_CLOSED.equals(status);
    }

    /** Copy with an edited title / description (the {@code record_pr_description}
     *  tool + the PATCH endpoint). A null argument leaves that field unchanged. */
    public LocalPR withDetails(String newTitle, String newDescription)
    {
        return new LocalPR(
                id, taskId, branchName, baseBranch,
                newTitle == null ? title : newTitle,
                newDescription == null ? description : newDescription,
                status, createdAt, pushedAt, remotePrNumber, remotePrUrl, mergedAt, closedAt,
                localAddressedThroughAt);
    }

    /**
     * Copy at a new status, stamping the terminal / push timestamps the flip
     * implies. Callers must have checked {@link #canTransitionTo}; the
     * timestamps ({@code when}, plus push metadata for a push flip) are set
     * only for the states that own them.
     */
    public LocalPR withStatus(String newStatus, Instant when)
    {
        return new LocalPR(
                id, taskId, branchName, baseBranch, title, description, newStatus, createdAt,
                STATUS_REMOTE_DRAFTED.equals(newStatus) && pushedAt == null ? when : pushedAt,
                remotePrNumber, remotePrUrl,
                STATUS_MERGED.equals(newStatus) ? when : mergedAt,
                STATUS_CLOSED.equals(newStatus) ? when : closedAt,
                localAddressedThroughAt);
    }

    /** Copy recording the remote PR identity assigned on push. */
    public LocalPR withRemote(int number, String url, Instant when)
    {
        return new LocalPR(
                id, taskId, branchName, baseBranch, title, description, status, createdAt,
                pushedAt == null ? when : pushedAt, number, url, mergedAt, closedAt,
                localAddressedThroughAt);
    }

    /** Copy with the local-addressing marker advanced to {@code through} — the
     *  high-water mark past which every {@code local_pr_comment} has already
     *  triggered (or been folded into) an addressing turn. Mirrors {@code
     *  TaskReviewMarkerStore} for the remote loop, scoped here since a
     *  {@code LocalPR} is already 1:1 with its task. */
    public LocalPR withLocalAddressedThrough(Instant through)
    {
        return new LocalPR(
                id, taskId, branchName, baseBranch, title, description, status, createdAt,
                pushedAt, remotePrNumber, remotePrUrl, mergedAt, closedAt, through);
    }
}
